File: lib/src/internal/serializer/button_settings_serializer.dart
import 'package:usercentrics_sdk/src/internal/internal.dart';
import 'package:usercentrics_sdk/src/model/button_settings.dart';

class ButtonSettingsSerializer {
  static dynamic serialize(ButtonSettings? value) {
    return value == null
        ? null
        : {
            'type': _serializeButtonType(value.type),
            'isAllCaps': value.isAllCaps,
            'fontAssetPath': value.fontAssetPath,
            'textSize': value.textSize,
            'textColor': ColorSerializer.serialize(value.textColor),
            'backgroundColor': ColorSerializer.serialize(value.backgroundColor),
            'cornerRadius': value.cornerRadius,
          };
  }

  static dynamic _serializeButtonType(ButtonType? value) {
    switch (value) {
      case null:
        return null;
      case ButtonType.acceptAll:
        return 'ACCEPT_ALL';
      case ButtonType.denyAll:
        return 'DENY_ALL';
      case ButtonType.more:
        return 'MORE';
      case ButtonType.save:
        return 'SAVE';
    }
  }
}

File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/ButtonSettingsSerializer.kt
package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.ButtonSettings
import com.usercentrics.sdk.ButtonType
import com.usercentrics.sdk.flutter.api.FlutterActivityProvider
import com.usercentrics.sdk.flutter.api.FlutterAssetsProvider

internal fun Any?.deserializeButtonSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): ButtonSettings? {
    if (this !is Map<*, *>) return null

    return ButtonSettings(
        type = this["type"]?.deserializeButtonType()!!,
        isAllCaps = this["isAllCaps"] as? Boolean,
        font = this["fontAssetPath"].deserializeTypeface(assetsProvider, activityProvider),
        textColor = (this["textColor"] as? String)?.deserializeColor(),
        backgroundColor = (this["backgroundColor"] as? String)?.deserializeColor(),
        cornerRadius = this["cornerRadius"] as? Int,
        textSizeInSp = (this["textSize"] as? Number)?.toFloat(),
    )
}

internal fun Any?.deserializeButtonType(): ButtonType? {
    if (this !is String) return null
    return ButtonType.valueOf(this)
}